src/compiler/file_uploader/index.ts

```diff
@@ -0,0 +1,187 @@
+    }
+}
+
+async function uploadDirectory(
+    srcDir: string,
+    destDir: string,
+    chunkSize: number,
+    actor: ActorSubclass
+) {
+    try {
+        const contents = await readdir(srcDir);
+        for (const name of contents) {
+            const srcPath = join(srcDir, name);
+            const destPath = join(destDir, name);
+            // Await each upload so the canister doesn't get overwhelmed by requests
+            await upload(srcPath, destPath, chunkSize, actor);
+        }
+    } catch (error) {
+        console.error(`Error reading directory: ${error}`);
+    }
+}
+
+async function uploadFile(
+    srcPath: Src,
+    destPath: Dest,
+    chunkSize: number,
+    actor: ActorSubclass
+) {
+    console.info(`uploadFile: Uploading ${srcPath} to ${destPath}`);
+    const uploadStartTime = process.hrtime.bigint();
+    const fileSize = (await stat(srcPath)).size;
+    for (let i = 0; i < fileSize; i += chunkSize) {
+        const fileStream = createReadStream(srcPath, {
+            start: i,
+            end: i + chunkSize - 1,
+            highWaterMark: chunkSize
+        });
+
+        let startIndex = i;
+        for await (const data of fileStream) {
+            await throttle();
+            console.info(
+                `uploadFile: ${srcPath} | ${bytesToHumanReadable(
+                    i + data.length
+                )} of ${bytesToHumanReadable(fileSize)}`
+            );
+            // Don't await here! Awaiting the agent will result in about a 4x increase in upload time.
+            // The above throttling is sufficient to manage the speed of uploads
+            actor
+                .upload_file_chunk(destPath, uploadStartTime, i, data, fileSize)
+                .catch((error) => {
+                    console.error(error);
+                });
+            startIndex += data.length;
+        }
+    }
+    console.info(`uploadFile: finished ${srcPath}`);
+}
+
+function bytesToHumanReadable(sizeInBytes: number): string {
+    const suffixes = ['B', 'KiB', 'MiB', 'GiB'];
+    // TODO it would be nice if we could figure out an algorithm as simple as
+    // TODO this that didn't involve a mutation
+    let size = sizeInBytes;
+
+    for (const suffix in suffixes) {
+        if (size < 1024.0) {
+            return `${size.toFixed(2)} ${suffixes[suffix]}`;
+        }
+        size /= 1024.0;
+    }
+
+    return `${size.toFixed(2)} ${suffixes[suffixes.length - 1]}`;
+}
+
+async function throttle() {
+    // We can only process about 4Mib per second. So if chunks are about
+    // 2 MiB or less then we can only send off two per second.
+    if (process.env.DFX_NETWORK === 'ic') {
+        await new Promise((resolve) => setTimeout(resolve, 2_000)); // Mainnet requires more throttling. We found 2_000 by trial and error
+    }
+    await new Promise((resolve) => setTimeout(resolve, 500)); // Should be 500 (ie 1 every 1/2 second or 2 every second)
+}
+
+async function getIdentity(
```

Comment:
We need to look into what other kinds of identities are possible, for example my demergent-labs-lastmjs-0 identity does not work

Reply:
https://github.com/demergent-labs/azle/issues/1719 This doesn't seem critical for first launch. It will go from no users being able to upload to some users being able to upload. So it seems like, at least for organizations sake, we can separate this into a different feature